### DISCUSSION TITLE
Is grid.grad=0 when we use mode="nearest" in grid_sample?

### DISCUSSION THREAD

**Discussion by BailiangJ** [Project-MONAI/MONAI#2881] [Q&A]
Hi,

I saw that in the 3D registration tutorial,

https://github.com/Project-MONAI/tutorials/blob/master/3d_registration/paired_lung_ct.ipynb

```
warp_layer = Warp().to(device)
pred_image = warp_layer(moving_image, ddf)
pred_label = warp_layer(moving_label, ddf)
```

the label is using the same warper/resampler as the image, which by default uses mode='bilinear' in torch.nn.functional.grid_sample.

Why shouldn't we use mode='nearest' here? Is it because if we use mode='nearest', the gradient wrt grid would be zero?

https://discuss.pytorch.org/t/does-the-nn-funtional-grid-sample-backwards-grad-to-grid/31736

Thanks in advance!

**Comment by wyli** [comment]
yes I think this is task dependent, linear works well in this case, could you help confirm @kate-sann5100 @YipengHu

**Reply by BailiangJ** [reply]
Hi, @wyli , thanks for your reply.

Using 'bilinear' here to labels warping would generate some values that don't belong to any classes at the boundary,

which I think would be a problem when we one-hot the warped label and compute the dice loss?

**Reply by YipengHu** [reply]
@BailiangJ 
- Re mode=`nearest`: not sure what you mean by "gradient would be zero". So long the `grid_sample` has automatic differentiation, which i think it does, the change of the grid sampling locations is likely to lead to the change of the loss (due to warped images/labels) - it is a case of non-zero gradient. The difference between `nearest` and `bilinear` is largely application-dependent. For example, when an intensity/based-based similarity is used, `bilinear`/`trilinear` is likely to work better.
- Re the boundary issue: it is true binary loss may be affected by the extrapolation value/method, so does a "soft dice loss". Then again, how significant this is is also dependent on the application. For example, some non-rigid registration that only warps ROIs nowhere near the image boundaries, such as neuroimages, would have little to none adverse effect.

**Reply by kate-sann5100** [reply]
I think @YipengHu had given a nice explanation. I just want to add that this particular implementation is because we are dealing with binary labels and calculating `DiceLoss` with `include_background` set to False. 
When dealing with categorical labels with more than 2 classes, we suggest the following procedures:

If to use 'bilinear' warping:
1. convert the single-channel label to one-hot encoding
2. perform 'bilinear' warping on the one-hot label
3. use `argmax` function to convert warped label back to single-channel with int values

If to use 'nearest' warping:
simply perform 'nearest ' warping on the single channel label would be enough.

Hope this would be helpful.

**Reply by BailiangJ** [reply]
@YipengHu @kate-sann5100 
Thank you for your generous reply. It is clear to me now that the `grid_sample` mode is application-dependent.

For using `mode='nearest'`, I would still like to have some discussions here.

When I do

`
pred_label = F.grid_sample(moving_label, grid=flow, mode='nearest', align_corners=True)`
`loss = diceloss(pred_label, fixed_label)`
`backward()
`

I found that the gradient with respect to `flow` would be all zero, which is also discussed here:
https://discuss.pytorch.org/t/does-the-nn-funtional-grid-sample-backwards-grad-to-grid/31736/4

If this is the case, then the dice loss has 0 gradients backpropagating to the network parameters if using `mode='nearest'` to warp the moving label.

So at first, I thought that the notebook using `mode='bilinear'` to warp the moving label is intended to avoid this situation.

I think this would be an interesting problem for people doing registration because it usually comes first to my mind that using nearest neighbor mode to resample labels and using dice loss as the auxiliary loss in addition to image similarity.

But I don't know should I post the question on the behavior of `grid_sample(mode='nearest')` here or on a pytorch community forum.

Thanks a lot.

**Reply by YipengHu** [reply]
did you actually get all-zero gradient, with a float-value `flow`?

**Reply by BailiangJ** [reply]
Yes, here is the experiment I did:
```
num_iteration = 5
fixed_label = 
moving_label = 
#initial dice score > 0.6

theta = torch.tensor([1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0], dtype=torch.float).to(fixed_label)
theta = Variable(theta, requires_grad=True)

optimizer = torch.optim.Adam([theta], lr=lr)

for iter in range(num_iteration):
        optimizer.zero_grad()
        flow = F.affine_grid(theta, moving_label.size(), align_corners=True)
        warp_moving = F.grid_sample(moving_label, grid=flow, mode="nearest", align_corners=True)

        warp_moving.retain_grad()
        flow.retain_grad()

        loss = torch.sum( (fixed_label - warp_moving)**2 ) #I use SSE loss here
        print(f"iter {iter}, loss: {loss}")
        loss.backward()

        print(torch.unique(warp_moving.grad))
        print(torch.unique(flow.grad))

        optimizer.step()
```

and the result I got is:
```
iter 0, loss: 22536.0
tensor([-2.,  0.,  2.])
tensor([0.])
iter 1, loss: 22536.0
tensor([-2.,  0.,  2.])
tensor([0.])
iter 2, loss: 22536.0
tensor([-2.,  0.,  2.])
tensor([0.])
iter 3, loss: 22536.0
tensor([-2.,  0.,  2.])
tensor([0.])
iter 4, loss: 22536.0
tensor([-2.,  0.,  2.])
tensor([0.])
```

the code was not doing the affine registration as I wanted (loss is not decreasing),
but when I change the `mode="bilinear"` , the gradient w.r.t `flow` is not all zero,
and the dice score could get to 0.97 (initial is around 0.66) after 200 iterations

**Reply by YipengHu** [reply]
it looks like it settles the argument - although i'm a bit surprised PyTorch didn't come up with some clever approximation for non-zero gradient - we should not only recommend but perhaps enforce the bilinear/trilinear mode. Thanks @BailiangJ 

@kate-sann5100 do you think you can reproduce this test and add a note/change to the PyTorch implementation accordingly?
@wyli does the MONAI cuda resampler has non-zero gradient with nearest option?

**Reply by kate-sann5100** [reply]
@BailiangJ Thank you for the useful information.

@YipengHu I will double-check the MONAI resampler gradient and amend the PyTorch implementation next week.

**Reply by YipengHu** [reply]
also, try initialise the matrix with some small random noise to see if it still holds.

**Reply by wyli** [reply]
in my understanding, it's zeros in the MONAI implementation as well. I'm not familiar with this, but it might be interesting to see some differentiable approximations? google returned this https://math.stackexchange.com/questions/842647/polynomial-equal-to-the-ceiling-of-x
